File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    answer = []
    tickets.sort()
    visited = [False] * len(tickets)
    
    def dfs(current_airport, path):
        # base 조건
        if len(path) == len(tickets)+1:
            return path
        
        for i, ticket in enumerate(tickets):
            if not visited[i] and ticket[0] == current_airport:
                visited[i]= True
                
                result = dfs(ticket[1], path + [ticket[1]])
                
                if result:
                    return result
                
                visited[i]= False
                
        return 0
    
    return dfs("ICN", ["ICN"])

File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;
vector<string> answer;
vector<bool> visited;
bool found; 

void dfs(string curr, vector<vector<string>>& tickets, vector<string>& path){
    if(found) return;
    if(path.size() == tickets.size()+1){
        answer=path;
        found = true;
        return;
    }
    for(int i=0; i<tickets.size(); i++){
        if(!visited[i]&&tickets[i][0]==curr){
            visited[i]=true;
            path.push_back(tickets[i][1]);
            dfs(tickets[i][1], tickets, path);
            path.pop_back();
            visited[i]=false;
        }
    }
}


vector<string> solution(vector<vector<string>> tickets) {
    sort(tickets.begin(), tickets.end());
    visited.assign(tickets.size(), false);
    vector<string> path;
    path.push_back("ICN");
    dfs("ICN",tickets, path);
    
    return answer;
}